### ISSUE TITLE
Improve asset-transfer-secured-agreement sample

### ISSUE THREAD

**Issue by denyeart** [hyperledger/fabric-samples#679]
This issue proposes to solve two issues in the existing [secured agreement asset transfer sample](https://github.com/hyperledger/fabric-samples/blob/main/asset-transfer-secured-agreement/chaincode-go/asset_transfer.go).

**ISSUE 1 - Single endorser for asset's state-based endorsement policy is vulnerable to bad actors**

Since only the current owner is required to endorse the transfer of an asset per the asset's state-based endorsement policy, it is possible for a malicious seller to use their own altered chaincode to transfer an asset in public state without writing the corresponding private state to the buyer's implicit private data collection. The solution to these types of bad actor scenarios when using Fabric is to require additional endorsers.

In terms of WHO the additional endorser should be, it depends on the use case but there are a several options. Depending on the use case, the state-based endorsement policy for an asset update could be:
1) The current owner AND any other channel member
2) The current owner AND the future buyer
3) The current owner AND a trusted third party such as a regulator.

Options 1) and 3) are possible in current Fabric.

Option 2) seems like a logical answer but would require an enhancement to Fabric state-based endorsement behavior. In existing Fabric when updating state-based endorsement policy for a key the transaction must be endorsed based on the outgoing endorsement policy for the key (seller in this sample). Fabric could potentially be enhanced for a namespace to optionally require that both the old and updated state-based endorsement policy be satisfied for a key (both seller and buyer endorse in this sample).
@ale-linux I believe this was considered when state-based endorsement was implemented, was there a specific reason why this option was not implemented? Do you agree it is a viable option?

Since this sample runs on test-network which only has two Orgs, we can hardcode the sample to require both Org1MSP (seller) and Org2MSP (buyer) in the state-based endorsement policy to get the point across that requiring multiple endorsers is a good idea, and document the various options for real world scenarios.

To keep the sample simple, we can also remove the ChangePublicDescription() aspect. In the sample this function was used for the seller to indicate that a public asset was for sale. But it wouldn't make sense to require Org1 and Org2 to both endorse this change.

**ISSUE 2 - Private data shouldn't be disclosed to all possible endorsers**

In use cases where the second endorser should be a trusted third party such as a regulator, you may not want to disclose the private data to the regulator. As such, instead of passing asset private data to the chaincode TransferAsset function via transient field (which then gets checked against the on-chain hash and written to buyer's private data collection), the buyer should pass the asset private data to the chaincode ahead of the transfer in the AgreeToBuy function. Both the agreed to price and asset private data would then be written to buyer's implicit collection. The TransferAsset function can then be updated to verify that the seller's and buyer's asset private data hash are identical, similar to how the existing sample verifies that the seller's and buyer's agreed to price are identical using GetPrivateDataHash() from their respective implicit collections. In this way the Transfer can be verified by a trusted third party who executes and endorses the chaincode, without ever disclosing the private data to them.

**Comment by denyeart** [comment]
@ale-linux Thanks. Concerning your 3rd point, the premise of the secured agreement sample is that seller and buyer must agree to some terms prior to the transfer. In this sample they do that by indicating agreement in their respective implicit private data collections with a hash going into public state. Then a 3rd party endorser such as regulator can execute/endorse the actual transfer. The actual transfer chaincode verifies several things, for example that the requester is indeed the current owner and that both parties have agreed to the same transfer terms (e.g. price, asset definition) by checking that their respective hashes match using GetPrivateDataHash() for each side. In this way the terms can remain private from the 3rd party.

**Comment by ale-linux** [comment]
I think that works well - provided that there be no systemic invariant to preserve (e.g. new "value" cannot be minted as part of an agreement, only transferred). In other words, the agreement is a local affair between initiator and responder. In that case, what you propose works well.

If the assumption doesn't hold however, I don't see how the correctness of those systemic invariants can be preserved if the regulator cannot see the cleartext inputs. If it can, all is good (except we've probably violated some confidentiality requirements). If it can't, then _maybe_ things could work if upon transfer, the recipient gets to see the chain of custody all the way back to minting, but that's also bad from a privacy/confidentiality perspective right?

**Comment by denyeart** [comment]
@ale-linux Another thing we can do is put the hash of the asset definition in the original public state. It would be set upon issuance and immutable. Each endorser including the 3rd party would also check that the agreed to hash for the current transfer matches the original hash at time of issuance. Clients can also query the history of the key to see the chain of custody and the original hash throughout the chain of custody, prior to submitting their agreement. I think that resolves the concern, right?

**Comment by denyeart** [comment]
To continue on the last comment, the hash of the asset definition can be the id in public state. This ensures that the definition can never change and makes it easy to verify the asset definition against the public id.

**Comment by fraVlaca** [comment]
I realized that hardcoding the two orgs as buyers and sellers at the creation of the asset for the state endorsement policy was probably not the optimal idea.

I was thinking: what if we set the state endorsing policy to both a specific buyer and seller only on the AgreeToSell function bypassing the seller org id in the parameters?

Consequently, I did modify the sample to set the sbe only to the owner at the creation of the asset and change it to both owner and seller once the owner calls AgreeToSell ( which it needs to call anyway before doing the actual transfer).

In this way not only we can set the state endorsing policy to the actual next buyer but, also, the next buyer does not have to endorse any transaction on updating the asset that comes before and including the AgreeToSell function(in this way I could also keep the ChangePublicDescription function)
